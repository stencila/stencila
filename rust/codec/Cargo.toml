[package]
name = "codec"
version = "0.0.0"
edition = "2021"

[features]
proptest-min = ["schema/proptest-min"]
proptest-low = ["schema/proptest-low"]
proptest-high = ["schema/proptest-high"]
proptest-max = ["schema/proptest-max"]

[dependencies]
codec-losses = { path = "../codec-losses" }
codec-mapping = { path = "../codec-mapping" }
node-strip = { path = "../node-strip" }
common = { path = "../common" }
format = { path = "../format" }
schema = { path = "../schema" }
status = { path = "../status" }
