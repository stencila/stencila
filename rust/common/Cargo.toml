[package]
name = "common"
version = "0.0.0"
edition = "2021"

[dependencies]
Inflector = "0.11.4"
async-recursion = "1.0.5"
async-trait = "0.1.77"
base64 = "0.21.7"
bs58 = "0.5.0"
chrono = { version = "0.4.34", features = ["serde"] }
clap = { version = "4.5.1", features = ["derive", "color"] }
derivative = "2.2.0"
derive_more = "0.99.17"
dirs = "5.0.1"
eyre = "0.6.12"
futures = "0.3.30"
glob = "0.3.1"
indexmap = { version = "2.2.3", features = ["serde"] }
itertools = "0.12.1"
maplit = "1.0.2"
once_cell = "1.19.0"
proc-macro2 = "1.0.78"
quote = "1.0.35"
regex = "1.10.3"
reqwest = { version = "0.11.24", features = ["json"] }
serde = { version = "1.0.197", features = ["derive", "rc"] }
serde_json = { version = "1.0.114", features = ["preserve_order"] }
serde_with = "3.6.1"
serde_yaml = "0.9.32"
similar = { version = "2.4.0", features = ["unicode"] }
slug = "0.1.5"
smart-default = "0.7.1"
smol_str = "0.2.1"
strum = { version = "0.26.1", features = ["derive"] }
syn = "2.0.51"
tar = "0.4.40"
tempfile = "3.10.1"
tokio = { version = "1.36.0", features = ["full", "tracing"] }
toml = "0.8.10"
tracing = { version = "0.1.40", features = ["log"] }
type-safe-id = "0.2.1"
uuid = "1.7.0"
xz2 = "0.1.7"
zip = "0.6.6"
