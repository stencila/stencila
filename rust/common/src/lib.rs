//! # Common dependencies
//!
//! This internal crate simply re-exports dependencies that are commonly used across other internal
//! crates. Rust has a small `std` library (["and that's OK"](https://blog.nindalf.com/posts/rust-stdlib/)).
//! This crate acts as our internal version of a standard library, like [`stdx`](https://github.com/brson/stdx)
//! and others.
//!
//! The primary benefit of this crate is that there is only one place that version numbers for
//! commonly used dependencies need to be updated. Some of these crates are in line to become
//! part of the `std` library (e.g. `once_cell`).

pub use async_recursion;
pub use async_trait;
pub use base64;
pub use bs58;
pub use chrono;
pub use clap;
pub use derivative;
pub use derive_more;
pub use dirs;
pub use eyre;
pub use futures;
pub use glob;
pub use indexmap;
pub use inflector;
pub use itertools;
pub use maplit;
pub use once_cell;
pub use proc_macro2;
pub use quote;
pub use rand;
pub use regex;
pub use reqwest;
pub use serde;
pub use serde_json;
pub use serde_with;
pub use serde_yaml;
pub use similar;
pub use slug;
pub use smart_default;
pub use smol_str;
pub use strum;
pub use syn;
pub use tar;
pub use tempfile;
pub use tokio;
pub use toml;
pub use tracing;
pub use type_safe_id;
pub use uuid;
pub use which;
pub use xz2;
pub use zip;
