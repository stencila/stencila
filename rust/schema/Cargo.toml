[package]
name = "schema"
version = "0.0.0"
edition = "2021"

[features]
proptest-min = ["proptest", "proptest-derive"]
proptest-low = ["proptest", "proptest-derive"]
proptest-high = ["proptest", "proptest-derive"]
proptest-max = ["proptest", "proptest-derive"]

[dependencies]
codec-dom-trait = { path = "../codec-dom-trait" }
codec-html-trait = { path = "../codec-html-trait" }
codec-jats-trait = { path = "../codec-jats-trait" }
codec-json5-trait = { path = "../codec-json5-trait" }
codec-losses = { path = "../codec-losses" }
codec-markdown-trait = { path = "../codec-markdown-trait" }
codec-text-trait = { path = "../codec-text-trait" }
common = { path = "../common" }
human_name = "2.0.3"
interim = { version = "0.1.1", features = ["chrono"] }
monostate = "0.1.11"
node-id = { path = "../node-id" }
node-patch-derive = { path = "../node-patch-derive" }
node-store = { path = "../node-store" }
node-strip = { path = "../node-strip" }
node-type = { path = "../node-type" }
node-walk-derive = { path = "../node-walk-derive" }
proptest = { version = "1.4.0", optional = true }
proptest-derive = { version = "0.4.0", optional = true }
serde_with = "3.7.0"

[dev-dependencies]
codec = { path = "../codec" }
codec-markdown = { path = "../codec-markdown" }
common-dev = { path = "../common-dev" }
node-authorship = { path = "../node-authorship" }
