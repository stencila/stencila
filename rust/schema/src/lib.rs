mod deserialize;
mod implem;
mod prelude;

#[rustfmt::skip]
mod types;
pub use types::*;

pub mod shortcuts;
pub mod transforms;
pub mod walk;

pub use node_id::NodeId;
pub use node_type::NodeType;

#[cfg(feature = "proptest")]
mod proptests;
