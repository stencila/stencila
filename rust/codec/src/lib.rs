use std::{collections::BTreeMap, path::Path};

use common::{
    async_trait::async_trait,
    eyre::{bail, Result},
    serde::{Deserialize, Serialize},
    serde_with::skip_serializing_none,
    smart_default::SmartDefault,
    strum::{Display, IntoEnumIterator},
    tokio::{
        fs::{create_dir_all, File},
        io::{AsyncReadExt, AsyncWriteExt},
    },
    tracing,
};
use format::Format;
use node_strip::StripScope;
use schema::{Node, NodeType};
use status::Status;

// Re-exports for the convenience of internal crates implementing `Codec`
pub use codec_losses::{Losses, LossesResponse};
pub use codec_mapping::{Mapping, MappingEntry};
pub use common;
pub use format;
pub use schema;
pub use status;

/// The direction of conversion
pub enum CodecDirection {
    Decode,
    Encode,
}

/// A codec for decoding/encoding between Stencila Schema nodes and alternative formats
#[async_trait]
pub trait Codec: Sync + Send {
    /// The name of the codec
    ///
    /// Used when listing codecs and to select a codec when the user specifies
    /// the relevant options on the command line e.g. `--to jats-pandoc`.
    /// Should be kebab-cased.
    fn name(&self) -> &str;

    /// The status of the codec
    ///
    /// Used when listing codecs and to warn users when using a codec that
    /// is not stable.
    fn status(&self) -> Status;

    /// The level of support that the codec provides for decoding from a format
    #[allow(unused)]
    fn supports_from_format(&self, format: &Format) -> CodecSupport {
        CodecSupport::None
    }

    /// The level of support that the codec provides for decoding from each format
    fn supports_from_formats(&self) -> BTreeMap<Format, CodecSupport> {
        Format::iter()
            .filter_map(|format| {
                let support = self.supports_from_format(&format);
                support.is_supported().then_some((format, support))
            })
            .collect()
    }

    /// The level of support that the codec provides for decoding for a [`NodeType`]
    #[allow(unused)]
    fn supports_from_type(&self, node_type: NodeType) -> CodecSupport {
        CodecSupport::None
    }

    /// The level of support that the codec provides for decoding for each [`NodeType`]
    fn supports_from_types(&self) -> BTreeMap<String, CodecSupport> {
        NodeType::iter()
            .filter_map(|node_type| {
                let support = self.supports_from_type(node_type);
                support
                    .is_supported()
                    .then_some((node_type.to_string(), support))
            })
            .collect()
    }

    /// Whether the codec supports decoding from bytes
    fn supports_from_bytes(&self) -> bool {
        false
    }

    /// Whether the codec supports decoding from string content
    fn supports_from_string(&self) -> bool {
        true
    }

    /// Whether the codec supports decoding from a file system path
    fn supports_from_path(&self) -> bool {
        true
    }

    /// The level of support that the codec provides for encoding to a format
    #[allow(unused)]
    fn supports_to_format(&self, format: &Format) -> CodecSupport {
        CodecSupport::None
    }

    /// The level of support that the codec provides for encoding to each format
    fn supports_to_formats(&self) -> BTreeMap<Format, CodecSupport> {
        Format::iter()
            .filter_map(|format| {
                let support = self.supports_to_format(&format);
                support.is_supported().then_some((format, support))
            })
            .collect()
    }

    /// The level of support that the codec provides for encoding for a [`NodeType`]
    #[allow(unused)]
    fn supports_to_type(&self, node_type: NodeType) -> CodecSupport {
        CodecSupport::None
    }

    /// The level of support that the codec provides for encoding for each [`NodeType`]
    fn supports_to_types(&self) -> BTreeMap<String, CodecSupport> {
        NodeType::iter()
            .filter_map(|node_type| {
                let support = self.supports_to_type(node_type);
                support
                    .is_supported()
                    .then_some((node_type.to_string(), support))
            })
            .collect()
    }

    /// Whether the codec supports encoding to bytes
    fn supports_to_bytes(&self) -> bool {
        false
    }

    /// Whether the codec supports encoding to string content
    fn supports_to_string(&self) -> bool {
        true
    }

    /// Whether the codec supports encoding to a file system path
    fn supports_to_path(&self) -> bool {
        true
    }

    /// Get a list of types that the codec has either lossy decoding, or encoding, or both
    fn lossy_types(&self, direction: Option<CodecDirection>) -> Vec<NodeType> {
        let mut types = Vec::new();

        for node_type in NodeType::iter() {
            if (direction.is_none() || matches!(direction, Some(CodecDirection::Decode)))
                && self.supports_from_type(node_type).is_lossy()
                && !types.contains(&node_type)
            {
                types.push(node_type)
            }

            if (direction.is_none() || matches!(direction, Some(CodecDirection::Encode)))
                && self.supports_to_type(node_type).is_lossy()
                && !types.contains(&node_type)
            {
                types.push(node_type)
            }
        }

        types
    }

    /// Whether the codec uses remote state
    ///
    /// Some formats (e.g. Google Docs) have their canonical state in a remote
    /// location (e.g. Google's servers) and Stencila only "mirrors" them in a local
    /// file containing an id or URL linking the file to the remote state.
    fn has_remote_state(&self) -> bool {
        false
    }

    /// Decode a Stencila Schema node from bytes
    #[allow(unused_variables, clippy::wrong_self_convention)]
    async fn from_bytes(
        &self,
        bytes: &[u8],
        options: Option<DecodeOptions>,
    ) -> Result<(Node, Losses)> {
        bail!(
            "Decoding from bytes is not implemented for codec `{}`",
            self.name()
        )
    }

    /// Decode a Stencila Schema node from a string
    #[allow(unused_variables, clippy::wrong_self_convention)]
    async fn from_str(&self, str: &str, options: Option<DecodeOptions>) -> Result<(Node, Losses)> {
        self.from_bytes(str.as_bytes(), options).await
    }

    /// Decode a Stencila Schema node from a file
    ///
    /// This function reads the file as a string and passes that on to `from_str`
    /// for decoding. If working with binary formats, you should override this function
    /// to read the file as bytes instead.
    #[tracing::instrument(skip(self, file))]
    async fn from_file(
        &self,
        file: &mut File,
        options: Option<DecodeOptions>,
    ) -> Result<(Node, Losses)> {
        if self.supports_from_bytes() {
            let mut content = Vec::new();
            file.read_to_end(&mut content).await?;
            self.from_bytes(&content, options).await
        } else {
            let mut content = String::new();
            file.read_to_string(&mut content).await?;
            self.from_str(&content, options).await
        }
    }

    /// Decode a Stencila Schema node from a file system path
    #[tracing::instrument(skip(self))]
    async fn from_path(
        &self,
        path: &Path,
        options: Option<DecodeOptions>,
    ) -> Result<(Node, Losses)> {
        if !path.exists() {
            bail!("Path `{}` does not exist", path.display());
        }

        let mut file = File::open(path).await?;
        self.from_file(&mut file, options).await
    }

    /// Encode a Stencila Schema node to bytes
    #[allow(unused_variables)]
    async fn to_bytes(
        &self,
        node: &Node,
        options: Option<EncodeOptions>,
    ) -> Result<(Vec<u8>, Losses)> {
        bail!(
            "Encoding to bytes is not implemented for codec `{}`",
            self.name()
        )
    }

    /// Encode a Stencila Schema node to a string
    #[allow(unused_variables)]
    async fn to_string(
        &self,
        node: &Node,
        options: Option<EncodeOptions>,
    ) -> Result<(String, Losses, Mapping)> {
        self.to_bytes(node, options).await.map(|(bytes, losses)| {
            (
                String::from_utf8_lossy(&bytes).to_string(),
                losses,
                Mapping::none(),
            )
        })
    }

    /// Encode a Stencila Schema to a file
    #[tracing::instrument(skip(self, node, file))]
    async fn to_file(
        &self,
        node: &Node,
        file: &mut File,
        options: Option<EncodeOptions>,
    ) -> Result<Losses> {
        let mut options = options.unwrap_or_default();
        if options.standalone.is_none() {
            options.standalone = Some(true);
        }

        let (content, losses) = if self.supports_to_bytes() {
            self.to_bytes(node, Some(options)).await
        } else {
            self.to_string(node, Some(options))
                .await
                .map(|(string, losses, ..)| (string.as_bytes().to_vec(), losses))
        }?;
        file.write_all(&content).await?;

        Ok(losses)
    }

    /// Encode a Stencila Schema to a file system path
    #[tracing::instrument(skip(self, node))]
    async fn to_path(
        &self,
        node: &Node,
        path: &Path,
        options: Option<EncodeOptions>,
    ) -> Result<Losses> {
        if let Some(parent) = path.parent() {
            create_dir_all(parent).await?;
        }
        let mut file = File::create(path).await?;
        self.to_file(node, &mut file, options).await
    }
}

/// The level of support that a codec provides for a format or node type
#[derive(Debug, Default, Display, Serialize)]
#[serde(crate = "common::serde")]
pub enum CodecSupport {
    #[default]
    None,
    HighLoss,
    LowLoss,
    NoLoss,
}

impl CodecSupport {
    /// Whether a format or node type is supported
    pub fn is_supported(&self) -> bool {
        !matches!(self, CodecSupport::None)
    }

    /// Whether there is any loss for a format or node type
    pub fn is_lossy(&self) -> bool {
        !matches!(self, CodecSupport::NoLoss)
    }
}

/// Decoding options
#[skip_serializing_none]
#[derive(Debug, SmartDefault, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "kebab-case", crate = "common::serde")]
pub struct DecodeOptions {
    /// The name of the codec to use for decoding
    ///
    /// If not supplied then the format will be used to choose a codec.
    pub codec: Option<String>,

    /// The format to be decode from
    ///
    /// Most codecs only decode one format. However, for those that handle multiple
    /// format it may be necessary to specify this option.
    pub format: Option<Format>,

    /// The media type to decode from
    ///
    /// In some cases (e.g. when decoding content from a HTTP response) the
    /// IANA Media Type (MIME) will be known or need to be specified.
    pub media_type: Option<String>,

    /// Scopes defining which properties of nodes should be stripped before decoding
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub strip_scopes: Vec<StripScope>,

    /// A list of node types to strip before decoding
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub strip_types: Vec<String>,

    /// A list of node properties to strip before decoding
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub strip_props: Vec<String>,

    /// The response to take when there are losses in the decoding
    #[default(_code = "LossesResponse::Warn")]
    pub losses: LossesResponse,
}

/// Encoding options
#[skip_serializing_none]
#[derive(Debug, SmartDefault, Clone, Serialize, Deserialize)]
#[serde(default, rename_all = "kebab-case", crate = "common::serde")]
pub struct EncodeOptions {
    /// The name of the codec to use for encoding
    ///
    /// If not supplied then the format will be used to choose a codec.
    pub codec: Option<String>,

    /// The format to encode to
    ///
    /// Most codecs only encode to one format. However, for those that handle multiple
    /// formats it may be necessary to specify this option.
    pub format: Option<Format>,

    /// Whether to encode as a standalone document
    ///
    /// Unless specified otherwise, this is the default when encoding to a file
    /// (as opposed to a string).
    pub standalone: Option<bool>,

    /// Whether to encode in compact form
    ///
    /// Some formats (e.g HTML and JSON) can be encoded in either compact
    /// or "pretty-printed" (e.g. indented) forms. If not specified, the default
    /// for the format will be used.
    pub compact: Option<bool>,

    /// Whether to encode HTML for use by Stencila's web themes, clients & components
    ///
    /// Adds ids and other attributes to HTML which provide compatibility
    /// with web themes, clients and components which operate on the browser DOM.
    pub dom: Option<bool>,

    /// Scopes defining which properties of nodes should be stripped before encoding
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub strip_scopes: Vec<StripScope>,

    /// A list of node types to strip before encoding
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub strip_types: Vec<String>,

    /// A list of node properties to strip before encoding
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub strip_props: Vec<String>,

    /// The response to take when there are losses in the encoding
    #[default(_code = "LossesResponse::Warn")]
    pub losses: LossesResponse,
}
