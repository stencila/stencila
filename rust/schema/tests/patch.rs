use std::{collections::BTreeMap, fs::read_to_string};

use codec::{format::Format, Codec, DecodeOptions};
use codec_markdown::MarkdownCodec;
use common::{eyre::Result, glob::glob, serde::Serialize, tokio};
use common_dev::{insta::assert_yaml_snapshot, pretty_assertions::assert_eq};
use node_authorship::author_roles;
use schema::{
    diff, patch,
    shortcuts::{art, p, sec, t},
    Article, Author, AuthorRole, Block, Figure, Inline, Node, NodeProperty, Paragraph, PatchNode,
    PatchOp, PatchPath, PatchSlot, PatchValue, Primitive, Strong, Text, TimeUnit, Visitor,
    WalkControl,
};

/// An individual fixture
#[derive(Debug, Serialize)]
#[serde(crate = "common::serde")]
struct Fixture {
    /// The old node read from the fixture file
    old: Node,

    /// The new node read from the fixture file
    new: Node,

    /// The operations required to go from old to new
    ops: Vec<(PatchPath, PatchOp)>,

    /// The number of author roles before applying ops
    authors_before: usize,

    /// The number of author roles after applying ops
    authors_after: usize,
}

/// A summary of all the fixtures in this test
/// This is snap shotted so we can catch changes in key metrics
#[derive(Debug, Serialize)]
#[serde(crate = "common::serde")]
struct Summary {
    /// The number of operations for each fixture to go from old to new
    ops_count: BTreeMap<String, usize>,

    /// The total number of operations across all fixtures
    ops_total: usize,

    /// The number of operations for each fixture to go from old to new
    author_diffs: BTreeMap<String, isize>,

    /// The total number of operations across all fixtures
    author_diffs_sum: isize,
}

/// A visitor to count the author roles in a node
struct Authors(usize);

impl Authors {
    fn count(&mut self, authors: &Option<Vec<Author>>) {
        self.0 += authors.as_ref().map(|vec| vec.len()).unwrap_or_default()
    }
}

impl Visitor for Authors {
    fn visit_inline(&mut self, inline: &Inline) -> WalkControl {
        macro_rules! apply {
            ($($variant:ident),*) => {
                use Inline::*;
                match inline {
                    $($variant(node) => self.count(&node.options.authors),)*
                    _ => {}
                }
            };
        }

        // Should be applied to all inlines with the `authors` property
        apply!(
            AudioObject,
            CodeExpression,
            CodeInline,
            ImageObject,
            InstructionInline,
            MathInline,
            MediaObject,
            StyledInline,
            VideoObject
        );

        WalkControl::Continue
    }

    fn visit_block(&mut self, block: &Block) -> WalkControl {
        macro_rules! apply {
            ($($variant:ident),*) => {
                use Block::*;
                match block {
                    $($variant(node) => self.count(&node.options.authors),)*
                    _ => {}
                }
            };
        }

        // Should be applied to all blocks with the `authors` property
        apply!(
            Admonition,
            Claim,
            CodeBlock,
            CodeChunk,
            Figure,
            ForBlock,
            Heading,
            IfBlock,
            InstructionBlock,
            List,
            MathBlock,
            Paragraph,
            QuoteBlock,
            StyledBlock,
            Table
        );

        WalkControl::Continue
    }
}

// Count the authors in a node
fn authors(node: &Node) -> usize {
    let mut authors = Authors(0);
    authors.visit(node);
    authors.0
}

/// Snapshot tests of the `MergeNode::diff` method
#[tokio::test]
async fn fixtures() -> Result<()> {
    let mut ops_count = BTreeMap::new();
    let mut ops_total = 0;
    let mut author_diffs = BTreeMap::new();
    let mut author_diffs_sum = 0;
    for path in glob("tests/fixtures/*.md")?.flatten() {
        let name = path
            .file_stem()
            .and_then(|name| name.to_str())
            .unwrap()
            .to_string();

        // Read in the fixture and split into old and new Markdown
        let content = read_to_string(path)?;
        let mut parts = content.splitn(2, "===\n").map(String::from);
        let (old, new) = (
            parts.next().unwrap_or_default(),
            parts.next().unwrap_or_default(),
        );

        // Decode the old and new Markdown into nodes
        let options = Some(DecodeOptions {
            format: Some(Format::Markdown),
            ..Default::default()
        });
        let codec = MarkdownCodec {};
        let (old, ..) = codec.from_str(&old, options.clone()).await?;
        let (new, ..) = codec.from_str(&new, options).await?;

        // To test the retention of properties not represented in source Markdown
        // apply the default author role recursively to the old node
        let mut enriched = old.clone();
        author_roles(&mut enriched, vec![AuthorRole::default()]);
        let authors_before = authors(&enriched);

        // Calculate the ops
        let ops = diff(&old, &new)?;
        ops_count.insert(name.clone(), ops.len());
        ops_total += ops.len();

        // Apply ops and assert that get new node
        let mut merged = old.clone();
        patch(&mut merged, ops.clone())?;
        assert_eq!(merged, new, "{name}\n{ops:#?}");

        // Apply the ops to enriched node and count authors after
        patch(&mut enriched, ops.clone())?;
        let authors_after = authors(&enriched);
        let author_diff = authors_after as isize - authors_before as isize;
        author_diffs.insert(name.clone(), author_diff);
        author_diffs_sum += author_diff;

        // Snapshot the fixture
        assert_yaml_snapshot!(
            name,
            Fixture {
                old,
                new,
                ops,
                authors_before,
                authors_after
            }
        );
    }

    // Snapshot summary
    assert_yaml_snapshot!(
        "summary",
        Summary {
            ops_count,
            ops_total,
            author_diffs,
            author_diffs_sum
        }
    );

    Ok(())
}

#[test]
fn atoms() -> Result<()> {
    // Boolean

    assert_eq!(diff(&true, &true)?, vec![]);

    let mut old = true;
    let new = false;
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Set(new.to_value()?))]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Integer

    assert_eq!(diff(&1_i64, &1_i64)?, vec![]);

    let mut old = 1_i64;
    let new = 2_i64;
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Set(new.to_value()?))]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Number

    assert_eq!(diff(&1_f64, &1_f64)?, vec![]);

    let mut old = 1_f64;
    let new = 2_f64;
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Set(new.to_value()?))]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // String

    assert_eq!(diff(&String::from("abc"), &String::from("abc"))?, vec![]);

    let mut old = String::from("abc");
    let new = String::from("bcd");
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Set(new.to_value()?))]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    Ok(())
}

#[test]
fn vecs() -> Result<()> {
    // No ops: Both empty
    let mut old: Vec<i32> = vec![];
    let new = vec![];
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Push: Old empty, new only has one
    let mut old = vec![];
    let new = vec![1];
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Push(1.to_value()?))]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Append: Old empty, new has more than one
    let mut old = vec![];
    let new = vec![1, 2, 3];
    let ops = diff(&old, &new)?;
    assert_eq!(
        ops,
        vec![(
            PatchPath::new(),
            PatchOp::Append(vec![1.to_value()?, 2.to_value()?, 3.to_value()?])
        )]
    );
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Clear: New empty
    let mut old = vec![1, 2, 3];
    let new = vec![];
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Clear)]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Remove: all different
    let mut old = vec![1, 2, 3, 4, 5, 6, 7];
    let new = vec![1, 3, 7];
    let ops = diff(&old, &new)?;
    assert_eq!(
        ops,
        vec![(PatchPath::new(), PatchOp::Remove(vec![1, 3, 4, 5]))]
    );
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Remove: some same
    let mut old = vec![1, 1, 2, 2, 2, 3, 3, 4, 5, 5];
    let new = vec![1, 2, 3, 3, 4, 5];
    let ops = diff(&old, &new)?;
    assert_eq!(
        ops,
        vec![(PatchPath::new(), PatchOp::Remove(vec![1, 3, 4, 9]))]
    );
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Push: adding one
    let mut old = vec![1];
    let new = vec![1, 2];
    let ops = diff(&old, &new)?;
    assert_eq!(ops, vec![(PatchPath::new(), PatchOp::Push(2.to_value()?))]);
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Append: adding more than one
    let mut old = vec![1];
    let new = vec![1, 1, 2, 3];
    let ops = diff(&old, &new)?;
    assert_eq!(
        ops,
        vec![(
            PatchPath::new(),
            PatchOp::Append(vec![1.to_value()?, 2.to_value()?, 3.to_value()?])
        )]
    );
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Insert
    let mut old = vec![1, 3];
    let new = vec![0, 1, 2, 3, 4, 5];
    let ops = diff(&old, &new)?;
    assert_eq!(
        ops,
        vec![(
            PatchPath::new(),
            PatchOp::Insert(vec![
                (0, 0.to_value()?),
                (2, 2.to_value()?),
                (4, 4.to_value()?),
                (5, 5.to_value()?)
            ])
        )]
    );
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    // Change: same size, all different
    let mut old = vec![1, 2];
    let new = vec![3, 4];
    let ops = diff(&old, &new)?;
    assert_eq!(
        ops,
        vec![
            (
                PatchPath::from([PatchSlot::Index(0)]),
                PatchOp::Set(3.to_value()?)
            ),
            (
                PatchPath::from([PatchSlot::Index(1)]),
                PatchOp::Set(4.to_value()?)
            )
        ]
    );
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    Ok(())
}

#[test]
fn vec_sec() -> Result<()> {
    // This is a regression test for a bug found during testing
    let mut old = art([sec([p([t("para1")])])]);
    let new = art([sec([p([t("para1")]), p([t("para2")])])]);
    let ops = diff(&old, &new)?;
    assert!(matches!(ops[0].1, PatchOp::Push(..)));
    patch(&mut old, ops)?;
    assert_eq!(old, new);

    Ok(())
}

#[test]
fn enums() -> Result<()> {
    // Equal variants and values: no ops

    let node = Node::Article(Article::default());
    assert_eq!(diff(&node, &node)?, vec![]);

    let block = Block::Paragraph(Paragraph::default());
    assert_eq!(diff(&block, &block)?, vec![]);

    let inline = Inline::Text(Text::default());
    assert_eq!(diff(&inline, &inline)?, vec![]);

    let primitive = Primitive::Integer(1);
    assert_eq!(diff(&primitive, &primitive)?, vec![]);

    let time_unit = TimeUnit::Millisecond;
    assert_eq!(diff(&time_unit, &time_unit)?, vec![]);

    // Different variants: single replace op at root

    let node1 = Node::Article(Article::default());
    let node2 = Node::Integer(1);
    assert_eq!(
        diff(&node1, &node2)?,
        vec![(PatchPath::new(), PatchOp::Set(PatchValue::Node(node2)))]
    );

    let block1 = Block::Paragraph(Paragraph::default());
    let block2 = Block::Figure(Figure::default());
    assert_eq!(
        diff(&block1, &block2)?,
        vec![(PatchPath::new(), PatchOp::Set(PatchValue::Block(block2)))]
    );

    let inline1 = Inline::Text(Text::default());
    let inline2 = Inline::Strong(Strong::default());
    assert_eq!(
        diff(&inline1, &inline2)?,
        vec![(PatchPath::new(), PatchOp::Set(PatchValue::Inline(inline2)))]
    );

    let primitive1 = Primitive::Integer(1);
    let primitive2 = Primitive::String(String::new());
    assert_eq!(
        diff(&primitive1, &primitive2)?,
        vec![(PatchPath::new(), PatchOp::Set(primitive2.to_value()?))]
    );

    let time_unit1 = TimeUnit::Day;
    let time_unit2 = TimeUnit::Month;
    assert_eq!(
        diff(&time_unit1, &time_unit2)?,
        vec![(PatchPath::new(), PatchOp::Set(time_unit2.to_value()?))]
    );

    // Same variants, different values: ops depend differences
    use PatchSlot::*;

    let node1 = art([]);
    let node2 = art([p([t("para1")])]);
    assert_eq!(
        diff(&node1, &node2)?,
        vec![(
            PatchPath::from([Property(NodeProperty::Content)]),
            PatchOp::Push(PatchValue::Block(p([t("para1")])))
        )]
    );

    Ok(())
}
