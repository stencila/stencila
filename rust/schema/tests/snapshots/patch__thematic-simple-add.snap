---
source: rust/schema/tests/patch.rs
expression: "Fixture { old, new, ops, authors_before, authors_after }"
---
old:
  type: Article
  content:
    - type: Paragraph
      content:
        - type: Text
          value: This is something.
    - type: Paragraph
      content:
        - type: Text
          value: And something else.
new:
  type: Article
  content:
    - type: Paragraph
      content:
        - type: Text
          value: This is something.
    - type: ThematicBreak
    - type: Paragraph
      content:
        - type: Text
          value: And something else.
ops:
  - - - content
    - Insert:
        - - 1
          - type: ThematicBreak
authors_before: 2
authors_after: 2
